codegen-v2/src/lib.rs

```diff
@@ -7,121 +7,59 @@
-            main_dir = parse_file(&child_path, main_dir)?;
-        } else {
-            panic!()
-        }
+impl From<YamlError> for Error {
+    fn from(err: YamlError) -> Self {
+        Error::YamlError(err)
     }
-
-    Ok(main_dir)
 }
 
-fn parse_file(path: &Path, mut main_dir: CHeaderDirectory) -> Result<CHeaderDirectory> {
-    if !path.is_file() {
-        return Err(Error::Todo);
+impl From<RenderError> for Error {
+    fn from(err: RenderError) -> Self {
+        Error::RenderError(err)
     }
+}
 
-    let parent = path.parent().ok_or(Error::Todo)?;
-
-    let content = std::fs::read_to_string(path).map_err(|_err| Error::Todo)?;
-    // TODO: Find a better solution for this.
-    let content: String = content.chars().filter(|c| c.is_ascii()).collect();
-
-    let mut reader = Reader::from(content.as_str());
-    let mut items = vec![];
-
-    loop {
-        let (pending, checked_out) = reader.checkout();
-        if let Ok(der) = GHeaderFileItem::derive(checked_out) {
-            let (derived, branch) = (der.derived, der.branch);
-            reader = pending.merge(branch);
-
-            if let GHeaderFileItem::HeaderInclude(include) = &derived {
-                let child_path = parent.join(Path::new(&include.0));
-
-                // Skip self-referencing includes.
-                if child_path == path {
-                    continue;
-                }
-
-                // If the `#include`d file is not yet parsed, parse it.
-                // Otherwise, skip it.
-                if !main_dir.map.contains_key(&child_path) {
-                    main_dir = parse_file(&child_path, main_dir)?;
-                }
-            }
-
-            // Indicates EOF, implying this loop must exit.
-            if let GHeaderFileItem::Eof = derived {
-                break;
-            }
-
-            items.push(derived);
-        } else {
-            reader = pending.discard();
-        }
+impl From<TemplateError> for Error {
+    fn from(err: TemplateError) -> Self {
+        Error::TemplateError(err)
     }
-
-    // Update the map with newly parsed items.
-    main_dir.map.insert(path.to_path_buf(), items);
-
-    Ok(main_dir)
 }
 
-#[test]
-#[ignore]
-fn test_parse_with_output() {
-    let path = Path::new("../include/");
-    let dir = parse(&path).unwrap();
+fn current_year() -> u64 {
+    use std::time::{SystemTime, UNIX_EPOCH};
 
-    println!("UNRECOGNIZED items:");
-    for (path, items) in dir.map {
-        println!("> ## FILE: {}", path.to_str().unwrap());
-        for item in items {
-            if let GHeaderFileItem::Unrecognized(item) = item {
-                println!("  > {}", item);
-            }
-        }
-    }
+    let now = SystemTime::now();
+    let seconds_since_epoch = now
+        .duration_since(UNIX_EPOCH)
+        .expect("System's time is set before the start of the Unix epoch");
+
+    // 31556952 seconds in a year
+    1970 + (seconds_since_epoch.as_secs() / 31556952)
```

Comment:
Does it count a leap year?
What do you think about using the [chrono](https://docs.rs/chrono/latest/chrono/) crate?
https://users.rust-lang.org/t/how-to-get-current-year-and-month-from-chrono-datetime/59689

Reply:
`chrono` is works well, but kind of overkill to pull it just for that. The seconds are just the average seconds including leap years, see https://www.rapidtables.com/calc/time/seconds-in-year.html (it does not have to bee this precise, tbh).